spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -860,6 +865,46 @@ trait OptimisticTransactionImpl extends TransactionalWrite
     }
   }
 
+  def reportAutoCompactStatsError(e: Throwable): Unit = {
+    recordDeltaEvent(deltaLog, "delta.collectStats", data = Map("message" -> e.getMessage))
+    logError(e.getMessage)
+  }
+
+  def collectAutoOptimizeStats(numAdd: Long, numRemove: Long, actions: Iterator[Action]): Unit = {
+    // Early exit if no files were added or removed.
+    if (numAdd == 0 && numRemove == 0) return
+    val collector = createAutoCompactStatsCollector()
+    if (collector.isInstanceOf[DisabledAutoCompactPartitionStatsCollector]) return
+    AutoCompactPartitionStats.instance(spark)
```

Comment:
Is `instance` needed?! Why not use `apply`?

Reply:
see https://github.com/delta-io/delta/pull/2414/files#r1446659873